build/can-autoplay.js

```
@@ -1,81 +1,75 @@
-const mp4Base64 = require('./mp4-base64')
-const mp3Base64 = require('./mp3-base64')
-
-function removeElement ($element) {
-  $element.parentElement.removeChild($element)
-}
-
-//
-// API
-//
-
-function video (opts) {
-  opts = Object.assign({ muted: false }, opts)
-
-  const $video = document.createElement('video')
-
-  if (opts.muted) {
-    $video.muted = true
-    $video.volume = 0
-  } else {
-    $video.muted = false
-    $video.volume = 0.0001
-  }
-  $video.src = mp4Base64
-  $video.style.position = 'absolute'
-  $video.style.width = 0
-  $video.style.height = 0
-  $video.style.pointerEvents = 'none'
-
-  return new Promise((resolve, reject) => {
-    $video.addEventListener('canplay', () => {
-      $video.play()
-        .then(() => resolve(true))
-        .catch(() => resolve(false))
-        .then(() => removeElement($video))
-    })
-
-    document.querySelector('body').appendChild($video)
-
-    setTimeout(() => resolve(false), 250)
+const media = require('./media')
+let error = null
+
+function audio (options) {
+  return startPlayback(setupDefaultValues(options), () => {
+    return {
+      element: document.createElement('audio'),
+      source: media.AUDIO
+    }
   })
 }
 
-function videoMuted () {
-  return video({ muted: true })
+function getError () {
```

That function will always return the last generated error.
Since we are using Promises, guess we should return the error as a `reject` argument.

---

Do you mean, `getError` should be promisified and it will always reject? 

```
return Promise.reject(error); 
```

Would not it be weird that in order to get latest error object, you have to `catch`?

```
canAutoplay.getError().catch(error => console.log(error))
```

---

Let's say I do this:

```js
canAutoplay.audio()
canAutoplay.video()
```

And both fail.

Now I can't get the error from each call.
The error, IMHO, is derived from the method call, even thought it's async (future).
Which makes sense to put it as a Promise return.

We can either reject the promise or resolve it passing the error as argument.

### Rejecting
```js
// Promise
canAutoplay.video()
  .then(() => {
    // auto-play supported
  })
  .catch(error => {
    // auto-play *not* supported
    console.log('Cant auto-play', error)
  })

// async/await
try {
  await canAutoplay.video()
} catch (error) {
  // auto-play *not* supported
}

```

### Returning error

---

We could do `resolve/reject`.

Other possible solutions: 

### Pass callback to a check, if you are interested in an error object?

```
canAutoplay
    .video({errorListener: error => console.log(error)})
    .then(result => console.log(result));
```

### Return Object

```
canAutoplay
    .video()
    then({success, error} => console.log(error));
```

---

I don't like them because the Promise API is designed to handle errors already.
And if we respect that we gain for free `async/await` with `try/catch` which, IMHO, feels more idiomatic.

I would go with:
```js
// Promise
canAutoplay.video()
  .then(() => console.log('success'))
  .catch(error => console.log('error', error))

// Same as
canAutoplay.video()
  .then(
    () => console.log('success'),
    error => console.log('error', error)
  )

// async/await
try {
  await canAutoplay.video()
  console.log('success')
} catch (error) {
  console.log('error', error)
}
```

---

My opinion is that a promise should reject on an error, which is something unexpected. In this case I am expecting valid output to be either can play or can't play. I wouldn't say can't play is an unexpected error :)

Also my suggestion would be to resolve with an object like { canPlay: false, error: error }

With es6 it's only a small different from a Boolean

canAutoplayVideo().then(({ canPlay } => ...)

---

Agree with that!

---

@NicolasSiver thoughts?

---

I think the most convenient way to use the library will be when it does not throw an error.
I will adjust API to return an object as a result + error.